tests/baselines/reference/assertionsAndNonReturningFunctions.types

```diff
@@ -55,7 +55,7 @@ function assertIsString(x) {
 */
 function assert2(check) {
 >assert2 : (check: boolean) => asserts check
->        : ^     ^^       ^^^^^             
+>        : ^     ^^       ^^^^^^^^^^^^^^^^^^

```

Comment:
How did this one end up regressing?

Reply:
This is a very interesting scoping issue. In js doc param and return tags, when you use a symbol you are actually referring to the symbol in the outer scope, even if there is a symbol in the parameter list that matches. This is fine in most cases, except for type predicates, where a symbol should refer to the parameter. Here is just some of the strangeness I found:


```ts
// @checkJs: true
const check =  "outer"
/**
 * @param {typeof check} check // this typeof check reffers to the outer variable
 * @returns {asserts check} /// this should reffer to the parameter? Although it's a bit confusing
*/
function assert2(check) {
    // Declaration is wrong here in 5.4
    // function assert2(check: typeof check): asserts check;
    if (!check) throw new Error();
}

/**
 * @param {typeof check} check
 * @returns {asserts check is typeof check} // So the check in assert should reffer to the parameter but the check in typeof should reffer to the outer variable?
*/
function assert3(check) {
    if (!check) throw new Error();
}

var c;
/**
 * @param {typeof check} check // this typeof check reffers to the outer variable
 * @returns {asserts c} /// You can assert an outer variable too! No error and is pulls it into the d.ts
*/
function assert4(check) {
    // Declaration is wrong here in 5.4
    // function assert2(check: typeof check): asserts check;
    if (!check) throw new Error();
}

assert4("outer")

```


I think a lot of this previously worked because the type printer took it on faith that the symbols in the return type annotation were checked before and were safe to use. This just points out the issue. I already have a follow on PR to fix this and bring back the reuse for the assert clause in this case, but I didn't want to pile on too much in this one.

Reply:
Wow, great example. Definitely a case I didn't think about when I introduced fake scopes.